mjpc/planners/cross_entropy/planner.h

```diff
@@ -0,0 +1,168 @@
+#include "mjpc/states/state.h"
+#include "mjpc/trajectory.h"
+
+namespace mjpc {
+
+class CrossEntropyPlanner : public RankedPlanner {
+ public:
+  // constructor
+  CrossEntropyPlanner() = default;
+
+  // destructor
+  ~CrossEntropyPlanner() override = default;
+
+  // ----- methods ----- //
+
+  // initialize data and settings
+  void Initialize(mjModel* model, const Task& task) override;
+
+  // allocate memory
+  void Allocate() override;
+
+  // reset memory to zeros
+  void Reset(int horizon,
+             const double* initial_repeated_action = nullptr) override;
+
+  // set state
+  void SetState(const State& state) override;
+
+  // optimize nominal policy using random sampling
+  void OptimizePolicy(int horizon, ThreadPool& pool) override;
+
+  // compute trajectory using nominal policy
+  void NominalTrajectory(int horizon, ThreadPool& pool) override;
+
+  // set action from policy
+  void ActionFromPolicy(double* action, const double* state, double time,
+                        bool use_previous = false) override;
+
+  // resample nominal policy
+  void UpdateNominalPolicy(int horizon);
+
+  // add noise to nominal policy
+  void AddNoiseToPolicy(int i);
+
+  // compute candidate trajectories
+  void Rollouts(int num_trajectory, int horizon, ThreadPool& pool);
+
+  // return trajectory with best total return
+  const Trajectory* BestTrajectory() override;
+
+  // visualize planner-specific traces
+  void Traces(mjvScene* scn) override;
+
+  // planner-specific GUI elements
+  void GUI(mjUI& ui) override;
+
+  // planner-specific plots
+  void Plots(mjvFigure* fig_planner, mjvFigure* fig_timer, int planner_shift,
+             int timer_shift, int planning, int* shift) override;
+
+  // optimizes policies, but rather than picking the best, generate up to
+  // ncandidates. returns number of candidates created.
+  int OptimizePolicyCandidates(int ncandidates, int horizon,
+                               ThreadPool& pool) override;
+  // returns the total return for the nth candidate (or another score to
+  // minimize)
+  double CandidateScore(int candidate) const override;
+
+  // set action from candidate policy
+  void ActionFromCandidatePolicy(double* action, int candidate,
+                                 const double* state, double time) override;
+
+  void CopyCandidateToPolicy(int candidate) override;
+
+  // ----- members ----- //
+  mjModel* model;
+  const Task* task;
+
+  // state
+  std::vector<double> state;
+  double time;
+  std::vector<double> mocap;
+  std::vector<double> userdata;
+
+  // policy
+  SamplingPolicy policy;  // (Guarded by mtx_)
+  SamplingPolicy candidate_policy[kMaxTrajectory];
+  SamplingPolicy previous_policy;
+
+  // scratch
+  std::vector<double> parameters_scratch;
+  std::vector<double> times_scratch;
+
+  // trajectories
+  Trajectory trajectory[kMaxTrajectory];
+  Trajectory nominal_traj;  // the nominal traj
+
+  mjData* nominal_data;  // dummy data for nominal traj
```

Comment:
new: added nominal traj and data for plotting purposes, since the policy doesn't use any of the rollouts, but rather averages the elite ones into a new one